Refactoring kubelet to separate object creation from initialization

vishh opened the pull request:
This PR is not expected to cause any functional changes.

Adding a WIP label since the full impact of this change is not clear. 

cc @yujuhong

vishh committed:
Refactoring kubelet to separate object creation from object logical initialization.

dchen1107 commented:
Let's not target this for 1.1

vishh commented:
This PR is necessary if we want to change the node condition to `NotReady`
if cgroups are not mounted.

On Fri, Sep 25, 2015 at 9:57 AM, Dawn Chen <notifications@github.com> wrote:

> Let's not target this for 1.1
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14542#issuecomment-143280787>
> .
>

yujuhong commented on pkg/kubelet/kubelet.go:
```diff
@@ -1875,14 +1872,9 @@ func (kl *Kubelet) syncLoop(updates <-chan PodUpdate, handler SyncHandler) {
-			continue
-		}
-		if !kl.doneNetworkConfigure() {
+		if rs := kl.runtimeState.errors(); len(rs) != 0 {
```
This seems semantically different from previous code because runtimeState would start with no errors. If syncNetworkStatus hasn't run even once, we'd assume the network is configured?

vishh replied:
Yes. Fixed.

yujuhong commented on pkg/kubelet/kubelet.go:
```diff
@@ -2217,24 +2207,21 @@ func (kl *Kubelet) recordNodeStatusEvent(event string) {
+			err = fmt.Errorf("Error on adding ip table rules: %v", err)
+			glog.Error(err)
 		}
 		if len(kl.podCIDR) == 0 {
```
`kl.podCIDR`, IIUC, is guarded by `kl.networkConfigMutex`. Remove the lock acquiring here may cause data race.

vishh replied:
Good catch. I meant to fix it yesterday night, but missed it.

yujuhong replied:
sigh....this part of the code is too error-prone :-(

yujuhong commented:
I like the direction of this PR, but think the problem goes beyond aggregating the initialization errors in a single place.
  - kubelet struct still contains variables accessed by multiple goroutines (kl.podCIDR). 
  - some of the code should be modularized moved out of kubelet.go so that the code is less error-prone.

As a reference, this was my attempt to isolate all api.Node related code to a separate file #14090 (which got reverted due to a deadlock), and a commit to fix the deadlock https://github.com/yujuhong/kubernetes/commit/f71bb6d0fb5b3fc3b853bda48598fbef2fdf4c37

vishh committed:
handling locks and update tests. Fail node if network is not initialized

vishh committed:
refactor podCIDR handling in kubelet to runtime state

vishh commented:
+1 to all your comments @yujuhong. As you mention, there are quite a few refactorings that needs to be done in kubelet. Do you think this PR is a step in the right direction? 

BTW I moved `podCIDR` handling as well.

yujuhong commented on pkg/kubelet/kubelet.go:
```diff
@@ -368,21 +364,19 @@ func NewMainKubelet(
 
-	go util.Until(klet.syncNetworkStatus, 30*time.Second, util.NeverStop)
-	if klet.kubeClient != nil {
-		// Start syncing node status immediately, this may set up things the runtime needs to run.
```
Is this not true anymore?

vishh replied:
I did not notice any cross dependencies between this logic and the rest of the function. The node status will still be synced before the sync loop starts.

yujuhong replied:
Hmm...I think it's like this:
* syncNodeStatus gets the podCIDR from the Node object, and sets it.
* syncNetworkStatus configures the cbr0 bridge based on podCIDR, and restart docker if necessary
* wait for runtime to be up

It's not very clear in the current code that this will always happen in the desired order through, so I think it may not happen in the first try, but will eventually succeed. 

Given that you are changing the order, it's the impossible that this may caused unexpected situation to happen: longer startup time or restarting docker much later (at syncPod). It may worth spend some extra time making sure this would not happen with your PR.

vishh replied:
As we discussed offline, this PR should not change the existing startup
sequence. I am running e2e a few times to validate my assertion.

On Fri, Sep 25, 2015 at 11:47 AM, Yu-Ju Hong <notifications@github.com>
wrote:

> In pkg/kubelet/kubelet.go
> <https://github.com/kubernetes/kubernetes/pull/14542#discussion_r40462969>
> :
>
> > @@ -368,21 +364,19 @@ func NewMainKubelet(
> >  	}
> >  	klet.containerManager = containerManager
> >
> > -	go util.Until(klet.syncNetworkStatus, 30*time.Second, util.NeverStop)
> > -	if klet.kubeClient != nil {
> > -		// Start syncing node status immediately, this may set up things the runtime needs to run.
>
> Hmm...I think it's like this:
>
>    - syncNodeStatus gets the podCIDR from the Node object, and sets it.
>    - syncNetworkStatus configures the cbr0 bridge based on podCIDR, and
>    restart docker if necessary
>    - wait for runtime to be up
>
> It's not very clear in the current code that this will always happen in
> the desired order through, so I think it may not happen in the first try,
> but will eventually succeed.
>
> Given that you are changing the order, it's the impossible that this may
> caused unexpected situation to happen: longer startup time or restarting
> docker much later (at syncPod). It may worth spend some extra time making
> sure this would not happen with your PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14542/files#r40462969>.
>

yujuhong replied:
It'd be greatly appreciated If you can add some comments about the startup sequence, so that the next person who touches the code would be aware of the potential issues.

vishh replied:
Tried to add some information. After this refactor, the start up requirements should be much easier to wrap your head around.

yujuhong commented:
> Do you think this PR is a step in the right direction?

I am good with the direction of this PR; was simply pointing out a few things that we should keep in mind :)

I would suggest multiple e2e runs to make sure everything's fine. I would also read again later to see if there is anything that I'd missed.

Thanks for refactoring!

vishh committed:
avoid requiring runtime to be up while initializing kubelet image manager

vishh commented:
@yujuhong: With some help some @dchen1107 I made the image manager initialization not depend on the container runtime to be up.

yujuhong commented:
lgtm sans a nit. shippable is not happy though.

vishh committed:
Move container manager into a separate package.
Inject container manager into Kubelet. This lets us stub out container
manager during integration testing.

vishh commented:
@yujuhong: I had to do a bit more refactoring to appease the integration tests. Overall, I think the refactoring is moving us in a positive direction. I'm interested in your comments.

yujuhong commented:
> @yujuhong: I had to do a bit more refactoring to appease the integration tests. Overall, I think the refactoring is moving us in a positive direction. I'm interested in your comments.

Sure. Let me know when it's ready for review. I probably won't get to it today though.

vishh commented:
@yujuhong: This PR is ready for review. Take your time to do the review. This has already waited a few weeks :)

karlkfi commented:
Compilation failed merged with master. Might need a rebase.

https://teamcity.mesosphere.io/viewLog.html?buildId=56480&buildTypeId=Oss_KubernetesMesos_5SmokeTestsDockerMesos&tab=buildLog&guest=1#_focus=442

```
[17:17:41][Step 4/4] _output/local/go/src/k8s.io/kubernetes/cmd/kubemark/hollow-node.go:134: not enough arguments in call to app.SimpleKubelet
```

yujuhong commented:
lgtm overall with some nits.

derekwaynecarr commented:
@vishh - this is a definite improvement.

vishh commented:
@derekwaynecarr: Acknowledged.

yujuhong commented:
lgtm overall but jenkins/shippable are not happy.

yujuhong commented:
@vishh , docker needs to be up before cadvisor.Start(). See the code below.
https://github.com/google/cadvisor/blob/master/manager/manager.go#L199
https://github.com/google/cadvisor/blob/master/container/docker/factory.go#L186

vishh commented:
Thanks @yujuhong. I will fix cAdvisor and then come back to this PR :)

On Mon, Oct 19, 2015 at 9:53 AM, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e build/test *failed* for commit 32d051c
> <https://github.com/kubernetes/kubernetes/commit/32d051ca90b8e9f549757e5f469a5bc0eb9ccc5e>
> .
>
>    - Build Log
>    <https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/32d051ca90b8e9f549757e5f469a5bc0eb9ccc5e/kubernetes-pull-build-test-e2e-gce/13245/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/32d051ca90b8e9f549757e5f469a5bc0eb9ccc5e/kubernetes-pull-build-test-e2e-gce/13245/_artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//13245>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14542#issuecomment-149278759>
> .
>

yujuhong commented:
@vishh, time to revive this PR? :-)

vishh commented:
Yes. I will get to this later this week.

On Mon, Nov 2, 2015 at 12:29 PM, Yu-Ju Hong <notifications@github.com>
wrote:

> @vishh <https://github.com/vishh>, time to revive this PR? :-)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14542#issuecomment-153147322>
> .
>

vishh committed:
Make cadvisor startup synchronous with container runtime initialization.
This is hopefully a temporary workaround.

Signed-off-by: Vishnu kannan <vishnuk@google.com>

vishh commented:
@yujuhong: Rebased this PR. It clears e2e hopefully.

vishh commented:
@k8s-bot: ok to test

vishh commented:
@k8s-bot: test this please
